internal/pkg/deploy/cloudformation/env.go

```diff
@@ -247,3 +225,99 @@ func (cf CloudFormation) upgradeEnvironment(in *deploy.CreateEnvironmentInput, t
 		return fmt.Errorf("update and wait for stack %s: %w", s.Name, err)
 	}
 }
+
+func (cf CloudFormation) environmentStack(env *deploy.CreateEnvironmentInput) (*cloudformation.Stack, error) {
+	bucketARN, err := arn.Parse(env.ArtifactBucketARN)
+	if err != nil {
+		return nil, err
+	}
+	stackConfig := stack.NewEnvStackConfig(env)
+	url, err := cf.uploadStackTemplateToS3(bucketARN.Resource, stackConfig)
+	if err != nil {
+		return nil, err
+	}
+	cfnStack, err := toStackFromS3(stackConfig, url)
+	if err != nil {
+		return nil, err
+	}
+	return cfnStack, nil
+}
+
+func (cf CloudFormation) waitAndDescribeStack(stackName string) (*cloudformation.StackDescription, error) {
+	var (
+		stackDescription *cloudformation.StackDescription
+		err              error
+	)
+	for {
+		stackDescription, err = cf.cfnClient.Describe(stackName)
+		if err != nil {
+			return nil, fmt.Errorf("describe stack %s: %w", stackName, err)
+		}
+
+		if cloudformation.StackStatus(aws.StringValue(stackDescription.StackStatus)).InProgress() {
+			// There is already an update happening to the environment stack.
+			// Best-effort try to wait for the existing update to be over before retrying.
+			_ = cf.cfnClient.WaitForUpdate(context.Background(), stackName)
+			continue
+		}
+		break
+	}
+	return stackDescription, err
+}
+
+// transformParameters removes or transforms each of the current parameters and does not add any new parameters.
+func (cf CloudFormation) transformParameters(
+	currParams []*awscfn.Parameter,
+	oldParams []*awscfn.Parameter,
+	transform func(new *awscfn.Parameter, old *awscfn.Parameter) *awscfn.Parameter) ([]*awscfn.Parameter, error) {
+
+	// Make a map out of `currParams` and out of `oldParams`.
+	curr := make(map[string]*awscfn.Parameter)
+	for _, p := range currParams {
+		curr[aws.StringValue(p.ParameterKey)] = p
+	}
+	old := make(map[string]*awscfn.Parameter)
+	for _, p := range oldParams {
+		old[aws.StringValue(p.ParameterKey)] = p
+	}
+
+	// Remove or transform each of the current parameters.
+	for k, p := range curr {
+		transformed := transform(p, old[k])
+		if transformed == nil {
+			delete(curr, k)
+		} else {
+			curr[k] = transformed
+		}
+	}
+
+	var params []*awscfn.Parameter
+	for _, param := range curr {
+		params = append(params, param)
+	}
+	return params, nil
+}
+
+func transformEnvControllerParameters(new *awscfn.Parameter, old *awscfn.Parameter) *awscfn.Parameter {
```

Comment:
What's the reason where we want to use the new param values for EnvControllerManaged parameters vs. `UsePreviousValue`? Can we add it as a comment to this function?

Reply:
Responding here but will add ⬇️  as the comment as well:

One of your suggested comment explain exactly when we want to "use the new param value".
> // The ParamKey doesn't exist in the old stack, use the new value.

Right after `env init`, the env stack's `Parameter` section doesn't have parameters such as `ALBWorkloads`. At the first time users run `env deploy` after `env init`, these parameters are "new parameters that do not exist in the old stack", in which case, we can't `UsePreviousValue`.

This is the only time we set the param value for env-controller-managed params from env stack. We want to `UsePreviousValue` otherwise.